### PR TITLE
[Search] Refactor Algolia search to support multiple indices

### PR DESCRIPTION
**Pull request by rkdavidson**
### 1. What does this PR do, or why is it needed?
The Apollos core package for Algolia search is coupled to the `ContentItem` type. There are several occurrences of "ContentItem" in [@apollosproject/data-connector-algolia-search](https://github.com/ApollosProject/apollos-apps/blob/master/packages/apollos-data-connector-algolia-search/src/data-source.js#L56).

That coupling prevents clean reuse of the Algolia search integrated `Search` dataSource for other data types, i.e. **Groups**.

This PR's goal is to add **Groups** index to Algolia, **without breaking schema changes**.
In order to do that, a CF re-implementation of the Apollos core Search was required so multiple Algolia indices can coexist via the API.

Algolia indices are still configured in `config.yml`, but the `ALGOLIA` values (and relevant `env` vars) are refactored [here](https://github.com/christfellowshipchurch/apollos-api/pull/197/files?file-filters%5B%5D=.js&file-filters%5B%5D=.yml&hide-deleted-files=true#diff-5a7db8dbadaef1b1b5a8738ba70b5ffac82b8e243732154165911284e08aad4bR55).

A new plain JS class [`SearchIndex`](https://github.com/christfellowshipchurch/apollos-api/pull/197/files#diff-e013475dec7ed142ec29c574110e1327fa906de52aa0715435ce9fa407519d63) abstracts the Algolia SDK calls. There is one instance per index — one for `Groups` and one for `ContentItems`.

Calls that used to look like this:
```javascript
this.context.dataSources.Search.byPaginatedQuery(/* ... */);
```
become name-spaced by index:
```javascript
this.context.dataSources.Search.index('ContentItems').byPaginatedQuery(/* ... */);
//                              ^^^^^^^^^^^^^^^^^^^^^^
```

**Details:**
- Logic coupled to `ContentItems` moved from `src/data/search` to a new base class for ContentItems [`src/data/content-item/search.js`](https://github.com/christfellowshipchurch/apollos-api/pull/197/files#diff-e01e80320932b5fb26d8b4a1b4dd1374170bbf27b1b9da038812291c463a3734)
- The `src/data/search` module is now a data agnostic replacement for [@apollosproject/apollos-data-connector-algolia-search](https://github.com/ApollosProject/apollos-apps/tree/master/packages/apollos-data-connector-algolia-search)
- `GroupItem` and `ContentItem` modules are responsible for indexing and searching their data in Algolia
- Algolia results conform to `type SearchResult`

### 2. What design trade-offs/decisions were made?
This is a pretty disruptive change to existing content item search, **but it should not break the schema!**

I'm confident functionality was replaced properly, but would appreciate a focused review on the new `SearchableContentItem` base class for ContentItems:
https://github.com/christfellowshipchurch/apollos-api/blob/f73f0172c09b29f19ea2f03d945917d5edfcd2d2/src/data/content-item/search.js#L98
https://github.com/christfellowshipchurch/apollos-api/blob/f73f0172c09b29f19ea2f03d945917d5edfcd2d2/src/data/content-item/data-source.js#L12

### 3. How do I test this PR?

Explore the following queries locally in graphql playground.
Open mobile app in simulator from `master` branch, and verify Discover tab functionality works.

But first....
**Update local `.env` values**
```
# Delete (Or, leave for now, until we can safely delete it later)
ALGOLIA_SEARCH_INDEX=ContentItems

# Add
ALGOLIA_SEARCH_INDEX_CONTENT_ITEMS=ContentItems
ALGOLIA_SEARCH_INDEX_GROUPS=Groups
```

**Searching for `Groups`**
```graphql
query searchGroups {
  searchGroups(query: {
    // text: "Celebrate"
    // campusNames: ["Royal Palm Beach"]
    // preferences: ["Crew (Men)"]
    // subPreferences: ["Study Together"]
  }, first: 20) {
    pageInfo {
      startCursor
      endCursor
    }
    edges {
      cursor
      title
      summary
      coverImage {
        sources {
          uri
        }
      }
      node {
        id
        ...groupSearchResultNodeFragment
      }
    }
  }
}

fragment groupSearchResultNodeFragment on Node {
  ... on Group {
    preference
    subPreference
    campus {
      name
    }
    dateTime {
      start
    }
  }
  ... on GroupItem {
    members: people(first: 1) {
      totalCount
    }
    leaders: people(first: 20, isLeader: true) {
      totalCount
      edges {
        node {
          id
          firstName
          lastName
          nickName
          photo {
            uri
          }
        }
      }
    }
  }
}
```

**Index a `Group` into Algolia**
_NOTE:_ Mutation accepts a global Apollos ID _or_ Rock entity ID
```graphql
mutation indexGroup {
  indexGroup(id: "Group:ABC123", action: "update", key: {ENV.APOLLOS_SECRET})
}
```

**Searching for `ContentItems`**
```graphql
query searchContent {
  search(query: "Christmas") {
    __typename
    pageInfo {
      startCursor
      endCursor
    }
    edges {
      __typename
      cursor
      title
      summary
      coverImage {
        sources {
          uri
        }
      }

      node {
        __typename
        id
        ... on MediaContentItem {
          audios {
            name
            sources {
              uri
            }
          }
        }
        ... on InformationalContentItem {
          callsToAction {
            call
            action
          }
        }
      }
    }
  }
}
```

**Index a `ContentItem` into Algolia**
```graphql
mutation indexContentItem {
  indexContentItem(id: "7839", action: "update", key: {ENV.APOLLOS_SECRET})
}
```

### 4. What automated tests did you write?
None

## SCREENSHOTS/GIFS

|Caption|Screenshot/GIF|
|---|---|
|GQL Playground: `Group` search||
|GQL Playground: `ContentItem` search ||
|Indexing `ContentItems`||
|Mobile App: Discover tab||
|Mobile App: Searching||
|Mobile App: Filtering||

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [x] No new warnings in tests, storybook, or apps
- [x] Assign relevant reviewers
- [x] Upload GIF(s) of iOS, Android, Web
- [ ] Get one peer approval before merging

## REVIEW:

**Manual QA**

- [x] QA branch on iOS and ensure it looks/behaves as expected
- [ ] QA branch on Android and ensure it looks/behaves as expected
- [x] QA branch on Web and ensure it looks/behaves as expected

**Code Review**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Review by jordanwade — APPROVED** [review]
👍 LGTM, checked this on iOS, web-app, web-app-v2 and everything seems to be working as expected. I'll leave it to @asleepingpanda for the final approval, just in case he sees any breaking changes. Nice Work @rkdavidson!

**Comment by rkdavidson** [comment]
Hey @asleepingpanda , have you had a chance to review the changes in this yet? Specifically around the `ContentItem` mechanisms and indexing stuff?

File: config.yml [hunk 1]
```diff
@@ -43,13 +43,30 @@ CHURCH_ONLINE:
...
   API_KEY: ${ALGOLIA_API_KEY}
-  # TODO: should we support multiple indices? how?
-  SEARCH_INDEX: ${CONTENT}_${ALGOLIA_SEARCH_INDEX}
-  CONFIGURATION:
-    searchableAttributes:
-      - title
-      - htmlContent
-      - unordered(summary)
+  INDICES:
+    # 🤔 Is it ok or acceptable to use TitleCase key names here?
```
[the remark this answers is not in the document]
Comment (asleepingpanda) [reply config.yml:47]: once we get into the `CONFIGURATION` object, we're good to go back to lower camel case

File: src/data/content-item/data-source.js [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
 import moment from 'moment-timezone';
 
 import { createVideoUrlFromGuid, getIdentifierType } from '../utils';
 
+import SearchableContentItem from './search';
+
 const { ROCK_MAPPINGS, ROCK, FEATURE_FLAGS } = ApollosConfig;
 
-export default class ContentItem extends coreContentItem.dataSource {
+export default class ContentItem extends SearchableContentItem {
```
Comment (asleepingpanda) [inline src/data/content-item/data-source.js:12]: Hmmm.... idk how much I like this convention.

it seems backwards since search is a higher functionality of the Content Item class. In this convention above, it's written as though Content Item is an extension of the base Searchable class. Is there a way to invert this relationship?

File: src/data/content-item/schema.js [hunk 3]
```diff
@@ -59,11 +59,16 @@ export default gql`
...
+  extend type Query {
     getContentItemByTitle(title: String!): ContentItem
     getCategoryByTitle(title: String!): ContentItem
     allEvents: [EventContentItem]
     featuredEvents: ContentItemsConnection
     sermons(first: Int, after: String): ContentItemsConnection
+    search(query: String!, first: Int, after: String): SearchResultsConnection
+  }
+
+  extend type Mutation {
```
Comment (asleepingpanda) [inline src/data/content-item/schema.js:71]: Yeah so I'm getting a bit nervous with the manageability of this. If a change is needed to be made for search, I wonder how well understood it would be to go into each unique data type to configure it's search abilities. As opposed to managing all of these Index types from the Search datasource directory

**Review by asleepingpanda — CHANGES_REQUESTED** [review]

File: config.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rkdavidson) [reply config.yml]: Fixed in commit 0bb6a0e 👍

File: src/data/content-item/data-source.js (hunk 2 above)
Comment (rkdavidson) [reply src/data/content-item/data-source.js]: Fixed in commits 0bb6a0e and f3a03e5 by just pulling all that functionality up into `src/data/content-item/data-source.js` where it largely was in the first place.

I liked the idea of hiding all that search-specific logic into a separate, self-contained place for clarity — but it did feel awkward too relying on a class inheritance chain.

There's probably further room to centralize some utils like clamping object byte size and keyword extraction, but no immediate value right this second IMO.

File: src/data/content-item/schema.js (hunk 3 above)
Comment (rkdavidson) [reply src/data/content-item/schema.js]: Fixed in commit 4b59956 , by consolidating the schema defs to the `src/data/search/schema.js` file instead.

Rambled about this on Slack, but, my rationale for keeping search-related schema changes local to the _things being searched_ is that the resolvers and data-source methods behind them are in the data types, anyway.

I.e. I can see from `search/schema.js` that groups are searchable, but in order to see how/where the implementation happens, I have to go over to `group-item/resolvers.js`. Again, possibly personal preference, but I find it confusing when things reach up and over into other things' cookie jars.

But if the pattern is inconsistent with Apollos core, I'd more strongly prefer to be consistent than match my subjective opinion.

I thought `CheckInable` works like my original approach, but I had it backwards in my head.

So does this commit 4b59956 address your concern?

File: src/data/search/schema.js [hunk 4]
```diff
@@ -1,15 +1,46 @@
...
 
-    extend type Mutation {
-        indexContentItem(id: String, action: INDEX_ACTION, key: String): String
-    }
+  enum INDEX_ACTION {
+    update
+    delete
+  }
+
+  input SearchGroupsInput {
```
Comment (asleepingpanda) [inline src/data/search/schema.js:22]: Update the search input to `SearchQueryInput` with fields: `key: String!` and `values: [String]`

**Review by asleepingpanda — CHANGES_REQUESTED** [review]
